cclib/method/bader.py

```diff
@@ -0,0 +1,204 @@
+import copy
+import random
+import numpy
+import logging
+import math
+
+from cclib.method.calculationmethod import Method
+from cclib.method.volume import electrondensity_spin
+from cclib.parser.utils import convertor
+
+# Distance between two adjacent grids (sqrt[2] or sqrt[3] for uniform Cartesian grid)
+_griddist = numpy.array(
+    [
+        [[1.73205, 1.41421, 1.73205], [1.41421, 1, 1.41421], [1.73205, 1.41421, 1.73205],],
+        [[1.41421, 1, 1.41421], [1, 1, 1], [1.41421, 1, 1.41421]],
+        [[1.73205, 1.41421, 1.73205], [1.41421, 1, 1.41421], [1.73205, 1.41421, 1.73205],],
+    ],
+    dtype=float,
+)
+
+
+class MissingInputError(Exception):
+    pass
+
+
+def __cartesian_dist(pt1, pt2):
+    """ Small utility function that calculates Euclidian distance between two points
+        pt1 and pt2 are numpy arrays representing a point in Cartesian coordinates """
+    return numpy.sqrt(numpy.einsum("ij,ij->j", pt1 - pt2, pt1 - pt2))
+
+
+class Bader(Method):
+    """Bader's QTAIM charges."""
+
+    # All of these are required for QTAIM charges.
+    required_attrs = ("homos", "mocoeffs", "nbasis", "gbasis")
+
+    def __init__(self, data, volume, progress=None, loglevel=logging.INFO, logname="Log"):
+        super(Bader, self).__init__(data, progress, loglevel, logname)
+
+        self.volume = volume
+        self.fragresults = None
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+
+        # First obtain charge densities on the grid
+        if len(self.data.mocoeffs) == 1:
+            self.chgdensity = electrondensity_spin(
+                self.data, self.volume, [self.data.mocoeffs[0][: self.data.homos[0]]]
+            )
+            self.chgdensity.data *= 2
+        else:
+            self.chgdensity = electrondensity_spin(
+                self.data,
+                self.volume,
+                [
+                    self.data.mocoeffs[0][: self.data.homos[0]],
+                    self.data.mocoeffs[1][: self.data.homos[1]],
+                ],
+            )
+
+        # Assign each grid point to Bader areas
+        self.fragresults = numpy.zeros(self.chgdensity.data.shape, "d")
+        next_index = 1
+
+        self.logger.info("Partitioning space into Bader areas.")
+
+        # Exclude the outermost positions
+        for xwalker in range(1, len(self.chgdensity.data) - 1):
```

Comment:
Like we discussed, consider decreasing the indentation. One way to achieve this is helper functions. Another is to use a generator for the triple loop. Likely other good ideas as well, and anything that decreases the indent will help readability.

Reply:
I've used a generator for the triple loop to reduce the indent by two. Thanks!